ephios/plugins/federation/views/api.py

```diff
@@ -28,3 +37,88 @@ def get_queryset(self):
+        challenge = oauth_client.create_code_challenge(verifier, "S256")
+        authorization_url, _ = oauth.authorization_url(
+            urljoin(guest.url, "api/oauth/authorize/"),
+            code_challenge=challenge,
+            code_challenge_method="S256",
+        )
+        return authorization_url
+
+    def _oauth_callback(self):
+        guest = get_object_or_404(FederatedGuest, pk=self.request.session["guest"])
+        oauth_client = WebApplicationClient(
+            client_id=guest.client_id, code_verifier=self.request.session["code_verifier"]
+        )
+        oauth = OAuth2Session(client=oauth_client)
+        token = oauth.fetch_token(
+            urljoin(guest.url, "api/oauth/token/"),
+            authorization_response=self.request.get_full_path(),
+            client_secret=guest.client_secret,
+            code_verifier=self.request.session["code_verifier"],
+        )
+        self.request.session["access_token"] = token["access_token"]
+        self.request.session.set_expiry(token["expires_in"])
+        user_data = requests.get(
+            urljoin(guest.url, "api/users/me/"),
+            headers={"Authorization": f"Bearer {token['access_token']}"},
+            timeout=5,
+        )
+        try:
+            user = FederatedUser.objects.get(
+                federated_instance=guest, email=user_data.json()["email"]
+            )
+        except FederatedUser.DoesNotExist:
+            user = FederatedUser.objects.create(
+                federated_instance=guest,
+                email=user_data.json()["email"],
+                first_name=user_data.json()["first_name"],
+                last_name=user_data.json()["last_name"],
+                date_of_birth=user_data.json()["date_of_birth"],
+            )
+            for qualification in user_data.json()["qualifications"]:
+                try:
+                    # Note that we assign the qualification on the host instance without further checks.
+                    # This may lead to incorrect qualifications if the inclusions for the qualification
+                    # are defined differently on the guest instance. We are accepting this as it should
+                    # not happen with the pre-defined qualifications as we are displaying a warning if
+                    # the user adapt these and custom qualifications will have different uuids anyway.
+                    user.qualifications.add(Qualification.objects.get(uuid=qualification["uuid"]))
+                except Qualification.DoesNotExist:
+                    for included_qualification in qualification["includes"]:
+                        try:
+                            user.qualifications.add(
+                                Qualification.objects.get(uuid=included_qualification["uuid"])
+                            )
+                        except Qualification.DoesNotExist:
+                            continue
+        self.request.session["federated_user"] = user.pk
```

Comment:
A qualification `includes` is only the direct includes, not the transitive ones, right? If so, I don't like that we only follow it for one layer here. It should be all including or None if the one couldn't be found. Only doing one round of breath first search might lead to behavior that confuses users.

Reply:
no, the serialization uses the results from collect_all_qualifications